packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskService.java

```diff
@@ -0,0 +1,704 @@
+
+          // Host unreachable? Unregister all tasks for that app.
+          unregisterAllTasksForAppId(appId);
+        }
+      }
+    });
+
+    sAppRecords.put(appId, appRecord);
+    return true;
+  }
+
+  private void invalidateAppRecord(String appId) {
+    AppRecordInterface appRecord = sAppRecords.get(appId);
+
+    if (appRecord != null) {
+      appRecord.invalidate();
+      sAppRecords.remove(appId);
+      sHeadlessTaskManagers.remove(appId);
+      Log.i(TAG, "Invalidated headless app '" + appId + "'.");
+    }
+  }
+
+  private void finishJobAfterTimeout(final JobService jobService, final JobParameters params, long timeout) {
+    new Timer().schedule(new TimerTask() {
+      @Override
+      public void run() {
+        jobService.jobFinished(params, false);
+      }
+    }, timeout);
+  }
+
+  public static Map<String, Object> jsonToMap(String jsonStr, boolean recursive) {
+    try {
+      return jsonToMap(new JSONObject(jsonStr), recursive);
+    } catch (JSONException e) {
+      return new HashMap<>();
+    }
+  }
+
+  private static Map<String, Object> jsonToMap(JSONObject json, boolean recursive) {
+    Map<String, Object> map = new HashMap<>();
+
+    try {
+      Iterator<?> keys = json.keys();
+
+      while (keys.hasNext()) {
+        String key = (String) keys.next();
+        Object value = json.get(key);
+
+        if (recursive) {
+          value = jsonObjectToObject(value, recursive);
+        }
+
+        map.put(key, value);
+      }
+    } catch (JSONException e) {
+      e.printStackTrace();
+    }
+    return map;
+  }
+
+  private static List<Object> jsonToList(JSONArray json, boolean recursive) {
+    List<Object> list = new ArrayList<>();
+
+    try {
+      for (int i = 0; i < json.length(); i++) {
+        Object value = json.get(i);
+
+        if (recursive) {
+          if (value instanceof JSONArray) {
+            value = jsonToList((JSONArray) value, recursive);
+          } else if (value instanceof JSONObject) {
+            value = jsonToMap((JSONObject) value, recursive);
+          }
+        }
+        list.add(value);
+      }
+    } catch (JSONException e) {
+      e.printStackTrace();
+    }
+    return list;
+  }
+
+  private static Object jsonObjectToObject(Object json, boolean recursive) {
+    if (json instanceof JSONObject) {
+      return jsonToMap((JSONObject) json, recursive);
+    }
+    if (json instanceof JSONArray) {
+      return jsonToList((JSONArray) json, recursive);
+    }
+    return json;
+  }
+
+  /**
+   *  Method that unversions class names, so we can always use unversioned task consumer classes.
+   */
+  private static String unversionedClassNameForClass(Class versionedClass) {
+    String className = versionedClass.getName();
+    return className.replaceFirst("\\^abi\\d+_\\d+_\\d+\\.", "");
+  }
+
+  /**
+   *  Returns unversioned class from versioned one.
+   */
+  private static Class unversionedClassForClass(Class versionedClass) {
```

Comment:
Yeah, eventually this could live in a binding or something, but since the task manager doesn't support vanilla RN apps, I don't think it's important right now. I have that on my todo list for making the module compatible with vanilla apps.